filecache/__init__.py

```diff
@@ -543,25 +597,88 @@ def _retrieve_from_s3(self, filename, local_path):
+
+        Raises:
+            FileNotFoundError: If the file does not exist.
+        """
+
+        filename = filename.replace('\\', '/').lstrip('/')
+        local_path = self._cache_root / filename
+
+        if not local_path.exists():
+            raise FileNotFoundError(f'File {local_path} does not exist')
+
+        if self._prefix_type == 'local':
+            # We don't do anything for local paths since the file is already in the
+            # correct location.
+            return
+
+        if self._logger:
+            self._logger.debug(
+                f'Uploading {local_path} to {self._full_prefix}{filename}')
+
+        self._upload_counter += 1
+
+        if self._prefix_type == 'gs':
+            return self._upload_to_gs(filename, local_path)
+
+        if self._prefix_type == 's3':
+            return self._upload_to_s3(filename, local_path)
+
+        raise AssertionError(
+            f'Internal error unknown prefix type {self._prefix_type}')  # pragma: no cover
+
+    def _upload_to_gs(self, filename, local_path):
+        blob_name = f'{self._prefix}{filename}'
+        blob = self._gs_bucket.blob(blob_name)
+
+        blob.upload_from_filename(str(local_path))
+
+    def _upload_to_s3(self, filename, local_path):
+        s3_key = f'{self._prefix}{filename}'
+
+        self._s3_client.upload_file(str(local_path), self._s3_bucket_name, s3_key)
+
     @contextlib.contextmanager
-    def open(self, filename, *args, **kwargs):
-        """Retrieve and open a file as a context manager using the same options as open().
+    def open(self, filename, mode='r', *args, **kwargs):
+        """Retrieve/open or open/upload a file as a context manager.
+
+        If `mode` is a read mode (like ``'r'`` or ``'rb'``) then the file will be first
+        retrieved by calling :meth:`retrieve` and then opened. If the `mode` is a write
+        mode (like ``'w'`` or ``'wb'``) then the file will be first opened for write, and
+        when this context manager is exited the file will be uploaded.
 
         Parameters:
             filename (str or Path): The filename to open.
 
         Returns:
-            file-like object: The same object as would be returned by the normal
-            open() function.
+            file-like object: The same object as would be returned by the normal `open()`
+            function.
         """
 
-        local_path = self.retrieve(filename)
-        with open(local_path, *args, **kwargs) as fp:
-            yield fp
+        if mode[0] == 'r':
+            local_path = self.retrieve(filename)
+            with open(local_path, mode, *args, **kwargs) as fp:
+                yield fp
+        else:  # 'w', 'x', 'a'
+            local_path = self.get_local_path(filename)
+            with open(local_path, mode, *args, **kwargs) as fp:
+                yield fp
+            self.upload(filename)
+
+    @property
+    def download_counter(self):
```

Comment:
<div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/SETI/rms-filecache/pull/4#issuecomment-2387454445>#95f4be</a><span> - 10/02/2024, 01:34 am</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The removal of the 'filecache' method might break existing code that relies on it.
</div>
<div id="fix">
<b>Fix:</b> Instead of removing the 'filecache' method, deprecate it and provide a migration path to the new API.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -569,3 +569,7 @@ class FileCachePrefix:
 +    def filecache(self):
 +        """Return the FileCache object associated with this instance.
 +        Deprecated: This method will be removed in a future version. Use the FileCache instance directly."""
 +        import warnings
 +        warnings.warn("The 'filecache' method is deprecated and will be removed in a future version. Use the FileCache instance directly.", DeprecationWarning, stacklevel=2)
 +        return self._filecache
```
</div>
</details>
</div>
<br>


</div>

Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged